File: moqt/internal/message/group.go
package message

import (
	"errors"
	"io"
)

type GroupMessage struct {
	SubscribeID   uint64
	GroupSequence uint64
}

func (g GroupMessage) Len() int {
	var l int

	l += VarintLen(uint64(g.SubscribeID))
	l += VarintLen(uint64(g.GroupSequence))

	return l
}

func (g GroupMessage) Encode(w io.Writer) error {
	msgLen := g.Len()
	b := make([]byte, 0, msgLen+VarintLen(uint64(msgLen)))

	b, _ = WriteMessageLength(b, uint64(msgLen))
	b, _ = WriteVarint(b, g.SubscribeID)
	b, _ = WriteVarint(b, g.GroupSequence)

	_, err := w.Write(b)

	return err
}

func (g *GroupMessage) Decode(src io.Reader) error {
	size, err := ReadMessageLength(src)
	if err != nil {
		return err
	}

	b := make([]byte, size)

	_, err = io.ReadFull(src, b)
	if err != nil {
		return err
	}

	num, n, err := ReadVarint(b)
	if err != nil {
		return err
	}
	g.SubscribeID = num
	b = b[n:]

	num, n, err = ReadVarint(b)
	if err != nil {
		return err
	}
	g.GroupSequence = num
	b = b[n:]

	if len(b) != 0 {
		return ErrMessageTooShort
	}

	return nil
}

var ErrMessageTooShort = errors.New("message too short")

File: moq-web/src/message/group.ts
import type { Writer, Reader } from "../internal/io";
import { varintLen } from "../internal/io";
import type * as internal from "../internal";

export interface GroupMessageInit {
    subscribeId?: bigint;
    sequence?: internal.GroupSequence;
}

export class GroupMessage {
    subscribeId: bigint;
    sequence: internal.GroupSequence;

    constructor(init: GroupMessageInit) {
        this.subscribeId = init.subscribeId ?? 0n;
        this.sequence = init.sequence ?? 0n;
    }

    get messageLength(): number {
        return varintLen(this.subscribeId) + varintLen(this.sequence);
    }

    async encode(writer: Writer): Promise<Error | undefined> {
        writer.writeVarint(this.messageLength);
        writer.writeBigVarint(this.subscribeId);
        writer.writeBigVarint(this.sequence);
        return await writer.flush();
    }

    async decode(reader: Reader): Promise<Error | undefined> {
        let [len, err] = await reader.readVarint();
        if (err) {
            return err;
        }

        [this.subscribeId, err] = await reader.readBigVarint();
        if (err) {
            return err;
        }

        [this.sequence, err] = await reader.readBigVarint();
        if (err) {
            return err;
        }

        if (len !== this.messageLength) {
            throw new Error(`message length mismatch: expected ${len}, got ${this.messageLength}`);
        }

        return undefined;
    }
}